core/src/main/scala/kafka/server/metadata/QuotaCache.scala

```diff
@@ -0,0 +1,296 @@
+   *
+   * @param filters       A collection of quota filters (entity type and a match clause).
+   * @param strict        True if we should only return entities which match all the filter clauses and have no
+   *                      additional unmatched parts.
+   * @return              A mapping of quota entities along with their quota values.
+   */
+  def describeClientQuotas(filters: Seq[ClientQuotaFilterComponent], strict: Boolean):
+      Map[ClientQuotaEntity, Map[String, Double]] = inReadLock(lock) {
+    describeClientQuotasInternal(filters, strict).map { case (entity, value) => convertEntity(entity) -> value}
+  }
+
+  // Visible for testing (QuotaEntity is nicer for assertions in test code)
+  private[metadata] def describeClientQuotasInternal(filters: Seq[ClientQuotaFilterComponent], strict: Boolean):
+      Map[QuotaEntity, Map[String, Double]] = inReadLock(lock) {
+
+    // Do some preliminary validation of the filter types and convert them to correct QuotaMatch type
+    val entityFilters: mutable.Map[String, QuotaMatch] = mutable.HashMap.empty
+    filters.foreach(component => {
+      val entityType = component.entityType()
+      if (entityFilters.contains(entityType)) {
+        throw new InvalidRequestException(s"Duplicate ${entityType} filter component entity type")
+      } else if (entityType.isEmpty) {
+        throw new InvalidRequestException("Unexpected empty filter component entity type")
+      } else if (!ClientQuotaEntity.isValidEntityType(entityType)) {
+        throw new UnsupportedVersionException(s"Custom entity type ${entityType} not supported")
+      }
+
+      // A present "match()" is an exact match on name, an absent "match()" is a match on the default entity,
+      // and a null "match()" is a match on the entity type
+      val entityMatch = if (component.`match`() != null && component.`match`().isPresent) {
+        ExactMatch(component.`match`().get())
+      } else if (component.`match`() != null) {
+        DefaultMatch
+      } else {
+        TypeMatch
+      }
+      entityFilters.put(entityType, entityMatch)
+    })
+
+    if (entityFilters.isEmpty) {
+      return Map.empty
+    }
+
+    // We do not allow IP filters to be combined with user or client filters
+    val matchingEntities: Set[QuotaEntity] = if (entityFilters.contains(ClientQuotaEntity.IP)) {
+      if (entityFilters.size > 1) {
+        throw new InvalidRequestException("Invalid entity filter component combination, IP filter component should " +
+          "not be used with user or clientId filter component.")
+      }
+      val ipMatch = entityFilters.get(ClientQuotaEntity.IP)
+      if (ipMatch.isDefined) {
+        ipMatch.get match {
+          case ExactMatch(ip) => ipEntityIndex.getOrElse(SpecificIp(ip), Set.empty).toSet
+          case DefaultMatch => ipEntityIndex.getOrElse(DefaultIp, Set.empty).toSet
+          case TypeMatch => ipEntityIndex.values.flatten.toSet
+        }
+      } else {
+        Set.empty
+      }
```

Comment:
A more idiomatic solution is:
```scala
ipMatch.fold(Set.empty[QuotaEntity]) {
  case ExactMatch(ip) => ...
  case DefaultMatch => ...
  case TypeMatch => ...
}
```
or
```scala
ipMatch match {

  case Some(ExactMatch(ip)) => ...
  case Som(DefaultMatch) => ...
  case Some(TypeMatch) => ...
  case None => Set.empty[QuotaEntity])
}
```
This comment applies to a few other places.

Reply:
Nice 👍